ledger/smt.go

```diff
@@ -0,0 +1,425 @@
+	}
+	s := &smt{
+		hash:              hash,
+		TrieHeight:        len(hash([]byte("height"))) * 8, // hash any string to get output length
+		retentionDuration: retentionDuration,
+	}
+	s.db = &cacheDB{
+		updatedNodes: byteCache{cache: updateCache},
+	}
+	s.loadDefaultHashes()
+	return s
+}
+
+// loadDefaultHashes creates the default hashes
+func (s *smt) loadDefaultHashes() {
+	s.defaultHashes = make([][]byte, s.TrieHeight+1)
+	s.defaultHashes[0] = defaultLeaf
+	var h []byte
+	for i := 1; i <= s.TrieHeight; i++ {
+		h = s.hash(s.defaultHashes[i-1], s.defaultHashes[i-1])
+		s.defaultHashes[i] = h
+	}
+}
+
+// Update adds a sorted list of keys and their values to the trie
+// If Update is called multiple times, only the state after the last update
+// is committed.
+// When calling Update multiple times without commit, make sure the
+// values of different keys are unique(hash contains the key for example)
+// otherwise some subtree may get overwritten with the wrong hash.
+func (s *smt) Update(keys, values [][]byte) ([]byte, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.atomicUpdate = true
+	ch := make(chan result, 1)
+	s.update(s.Root, keys, values, nil, 0, s.TrieHeight, false, true, ch)
+	result := <-ch
+	if result.err != nil {
+		return nil, result.err
+	}
+	if len(result.update) != 0 {
+		s.Root = result.update[:hashLength]
+	} else {
+		s.Root = nil
+	}
+
+	return s.Root, nil
+}
+
+// result is used to contain the result of goroutines and is sent through a channel.
+type result struct {
+	update []byte
+	err    error
+}
+
+// update adds a sorted list of keys and their values to the trie.
+// It returns the root of the updated tree.
+func (s *smt) update(root []byte, keys, values, batch [][]byte, iBatch, height int, shortcut, store bool, ch chan<- result) {
+	if height == 0 {
+		if bytes.Equal(values[0], defaultLeaf) {
+			ch <- result{nil, nil}
+		} else {
+			ch <- result{values[0], nil}
+		}
+		return
+	}
+	batch, iBatch, lnode, rnode, isShortcut, err := s.loadChildren(root, height, iBatch, batch)
+	if err != nil {
+		ch <- result{nil, err}
+		return
+	}
+	if isShortcut {
+		keys, values = s.maybeAddShortcutToKV(keys, values, lnode[:hashLength], rnode[:hashLength])
+		// The shortcut node was added to keys and values so consider this subtree default.
+		lnode, rnode = nil, nil
+		// update in the batch (set key, value to default to the next loadChildren is correct)
+		batch[2*iBatch+1] = nil
+		batch[2*iBatch+2] = nil
+	}
+
+	// Split the keys array so each branch can be updated in parallel
+	// Does this require that keys are sorted?  Yes, see Update()
+	lkeys, rkeys := s.splitKeys(keys, s.TrieHeight-height)
+	splitIndex := len(lkeys)
+	lvalues, rvalues := values[:splitIndex], values[splitIndex:]
+
+	if shortcut {
+		store = false    //stop storing only after the shortcut node.
+		shortcut = false // remove shortcut node flag
+	}
+	if (len(lnode) == 0) && (len(rnode) == 0) && (len(keys) == 1) && store {
```

Comment:
Remove excess parens
```suggestion
	if len(lnode) == 0 && len(rnode) == 0 && len(keys) == 1 && store {
```